src/reference/mod.rs

```diff
@@ -0,0 +1,404 @@
+//! that are available, but you should double-check the documentation of whatever graph DSL tool
+//! you're using to make sure.
+//!
+//! [standard library]: functions/index.html
+//!
+//! # Graph nodes
+//!
+//! You can use this graph DSL to create any graph structure that you want.  There are no
+//! limitations on which graph nodes you create, nor on how you use edges to connect the graph
+//! nodes.  You are not limited to creating a tree, and in particular, you are not limited to
+//! creating a tree that "lines" up with the parsed syntax tree.
+//!
+//! You use the [`node`][] function to create a new graph node.  The function returns a reference
+//! to the new node.  You will almost always assign that result to a variable so that you can then
+//! do interesting things with the node:
+//!
+//! [`node`]: functions/index.html#node
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!   let new_node = (node)
+//! }
+//! ```
+//!
+//! Graph nodes are usually attached to syntax nodes using [scoped variables](#variables), so that
+//! you can refer to them from multiple stanzas:
+//!
+//! ``` tsg
+//! (identifier) @id
+//! {
+//!   let @id::node = (node)
+//! }
+//!
+//! (dotted_name (identifier) @dotted_element)
+//! {
+//!   ; We will learn more about the attr statement below
+//!   attr (@dotted_element::node) kind = "dotted"
+//! }
+//! ```
+//!
+//! In this example, we will create a graph node for _every_ `identifier` syntax node.  Each of
+//! those syntax nodes will have a `node` scoped variable, containing a reference to its graph
+//! node.  But only the graph nodes of those identifiers that appear inside of a `dotted_name` will
+//! have a `kind` attribute.  And even though we used different capture names, inside of different
+//! queries, to find those `identifier` nodes, the graph node references in both stanzas refer to
+//! the same graph nodes.
+//!
+//! # Edges
+//!
+//! Edges are created via an `edge` statement, which specifies the two graph nodes that should be
+//! connected.  Edges are directed, and the `->` arrow in the `edge` statement indicates the
+//! direction of the edge.
+//!
+//! ``` tsg
+//! (import_statement name: (_) @name)
+//! {
+//!   let @name::source = (node)
+//!   let @name::sink = (node)
+//!   edge @name::source -> @name::sink
+//! }
+//! ```
+//!
+//! There can be at most one edge connecting any particular source and sink graph node in the
+//! graph.  If multiple stanzas create edges between the same graph nodes, those are "collapsed"
+//! into a single edge.
+//!
+//! # Attributes
+//!
+//! Graph nodes and edges have an associated set of **_attributes_**.  Each attribute has a name
+//! (which is an identifier), and a value.
+//!
+//! You add attributes to a graph node or edge using an `attr` statement:
+//!
+//! ``` tsg
+//! (import_statement name: (_) @name)
+//! {
+//!   let @name::source = (node)
+//!   let @name::sink = (node)
```

Comment:
Small nit: it looks like these graph nodes were declared in the previous stanza. Are we wanting this stanza to be independent of the previous one?

Reply:
Yeah in this case I had tried to make the stanza independent, but I hear your point that it's not entirely clear whether that's true of all the stanzas in the doc, or not.

Reply:
I think @rewinfrey highlights a salient point here. If `(node)` creates a new node, what does the following code do?
```
(identifier) @id
{
  let @id::node = (node)
}

(identifier) @id
{
  let @id::node = (node)
}
```
- Is it an error? And will this be flagged at runtime -- when the second assignment is attempted -- or at "compile time"?
- Does it silently reuse the same node for both `let`s? (Thus preserving the notion of immutability that `let` implies?)
- Does it create two distinct `node` nodes for each `identifier`? (I mean, almost certainly not. That seems like a terrible idea.)

If the `let @id::node = (node)` declaration can appear in multiple different stanzas, then this creates the possibility that a typo creates a completely unrelated node (which may not be apparent until things start breaking).

If the `let @id::node = (node)` declaration can appear in at most one stanza (for a given scoped variable), then this creates a somewhat awkward situation where one stanza becomes responsible for ensuring these nodes are created (and possibly creates some edges and attributes between these), and all other stanzas do not. (Which means if you want to get rid of that stanza, you have to make sure to push the declarations into another appropriate stanza lest everything breaks.)

Perhaps what is needed is a way to say "these scoped variables are available on these `tree-sitter` nodes", so above we would say `identifier` can have a `node` scoped variable. Then if we write `node @id::node`, we're actually saying "oh, and please make sure to emit a graph node for this scoped variable", but crucially it'll be the same graph node across all accesses. And typos should be prevented, since we can check each scoped variable access against the set of scoped variables defined on its corresponding `tree-sitter` node.

(This would also make it easy to see exactly what scoped variables are defined for a given `tree-sitter` node. Currently you would have to find all captures that mention that node and look in the associated stanza to see what scoped variables are set on it.)

I'm not entirely sure how this would interact with mutability, though.

Reply:
>     * Is it an error? And will this be flagged at runtime -- when the second assignment is attempted -- or at "compile time"?

This is the current semantics, and it doesn't matter whether it's a new node reference that you're assigning to the variable. In general, you can only assign to an immutable `let` variable once, across all of the stanzas of the file.  So this would be an error, as well:

```
(identifier) @id
{
  let @id::count = 9;
}

(identifier) @id
{
  ; This is invalid
  let @id::count = 17;
}
```

There's a similar story with mutable variables: you can only declare them with `var` once, and you can't `set` them unless you've already declared them.  Although the `var` declaration and `set` reassignment don't have to be in the same stanza:

```
; This is valid

(identifier) @id
{
  var @id::count = 9;
}

(identifier) @id
{
  set @id::count = 17;
}
```